test/tests/functional/pbs_offline_vnodes.py

```diff
@@ -0,0 +1,204 @@
+        self.server.expect(JOB, {ATTR_state: 'Q'}, id=jid)
+
+        self.server.expect(
+            NODE, {ATTR_NODE_state: 'offline'},
+            id=self.moms.values()[0].shortname)
+        if self.is_cray is True:
+            vlist = []
+            vnl = self.server.filter(
+                VNODE, {'resources_available.vntype': 'cray_compute'})
+            vlist = vnl["resources_available.vntype=cray_compute"]
+            # Loop through each compute vnode in the list and check if state =
+            # offline
+            for v1 in vlist:
+                # Check that the vnode is in offline state
+                self.server.expect(
+                    VNODE, {'state': 'offline'}, id=v1, max_attempts=3,
+                    interval=2)
+
+        else:
+            self.server.expect(
+                NODE, {ATTR_NODE_state: 'offline'}, id='vnode[0]')
+            self.server.expect(
+                NODE, {ATTR_NODE_state: 'offline'}, id='vnode[1]')
+
+    def test_multi_mom_hook_failure_affects_vnode(self):
+        """
+        Run an execjob_begin hook that sleeps for sometime,
+        at the same time set an alarm value so less that
+        the hook alarms out and server executes the fail_action
+        After this check if vnodes are marked offline.
+        In case of a multiple mom reporting same set of vnodes,
+        it should not mark the vnodes and mom as offline because
+        there are other moms active and reporting same vnodes.
+        """
+        self.create_multi_vnodes(2, False)
+        self.create_mom_hook()
+
+        # set one  natural node to have higher ncpus than the other one
+        self.server.manager(MGR_CMD_SET, NODE, {
+                            ATTR_rescavail + '.ncpus': '256'},
+                            id=self.moms.values()[0].shortname)
+        self.server.manager(MGR_CMD_SET, NODE, {
+                            ATTR_rescavail + '.ncpus': '1'},
+                            id=self.moms.values()[1].shortname)
+
+        j1 = Job(TEST_USER)
+        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
+                           id=self.moms.values()[0].shortname)
+        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
+                           id=self.moms.values()[1].shortname)
+
+        if self.is_cray is True:
+            # on a cray, make sure job runs on login node with higher number of
+            # ncpus
+            j1.set_attributes(
+                {ATTR_l + '.select': '1:ncpus=256:vntype=cray_login'})
+        else:
+            j1.set_attributes({ATTR_l + '.select': '1:ncpus=256'})
+
+        jid = self.server.submit(j1)
+        self.server.expect(JOB, {ATTR_state: 'Q'}, id=jid)
+
+        self.server.expect(
+            NODE, {ATTR_NODE_state: 'offline'},
+            id=self.moms.values()[0].shortname)
+        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
+                           id=self.moms.values()[1].shortname)

```

Comment:
Sometimes when you check for the node/vnode state you set max_attempts and an interval, other times you don't.  Is there a reason?  I think you should check a few times in all places, just in case.
Here it might be because you don't expect the state of mom1 to change...but just in case I think the test should wait a little to see if it does change.

Reply:
Ideally, it should not take more than a second or so to update the state that is why added max_attempts and interval in a few places. I'll update all other instances.